woocommerce/woocommerce-android
# Issue/2709 Update downloadable files for a product

Fixes #2709

**Description**
This PR adds the ability to update downloadable files of a product:
1. Changing the file details: name and url
2. Changing the download limit, and download expiry
3. Reordering the files using drag and drop

**Changes**

1. Added ProductDownloadDetailsFragment with its own viewmodel to handle the update. The naming doesn't mention Edit specifically because the goal is to reuse the same fragment for adding files too.
2. Added ProductDownloadDetailsViewModelTests to test the different scenarios handled by the new fragment.
3. Instead of returned the result to the parent fragment, I opted to use the parent viewmodel directly (navgraph scoped) as it's easier to read, and avoids the extra layer of request codes:
https://github.com/woocommerce/woocommerce-android/blob/1e44e63b11de7671e6519770633cb81bf4a31327/WooCommerce/src/main/kotlin/com/woocommerce/android/ui/products/downloads/ProductDownloadDetailsFragment.kt#L36
4. Because of having multi-level navigation in this feature, there is an known issue with the "discard changes" dialog when coming back from the editing the details of a file, it's tracked here: #2737

**Screenshots**
<img width="347" alt="Screen Shot 2020-08-18 at 08 32 17" src="https://user-images.githubusercontent.com/1657201/90483911-60adc400-e12d-11ea-87da-61d635f2b556.png">
<img width="347" alt="Screen Shot 2020-08-18 at 08 32 00" src="https://user-images.githubusercontent.com/1657201/90483913-62778780-e12d-11ea-871f-4623e7e858aa.png">
![update-downloads](https://user-images.githubusercontent.com/1657201/90339911-30aad780-dfec-11ea-9d1d-5f588eaceb25.gif)


**Testing**
1. Make sure a product in the store is both `downloadable` and has some `downloadable files` (more than one file).
2. Go to the product details on the app.
3. Click on "Downloadable files" section.
4. Test reordering the files by clicking on the handle and dragging.
5. Click on one of the files and test editing its details.
6. Click on the settings button in the files list, and edit the "download limit" and "download expiry"
